src/lib/y2storage/proposal/space_maker_prospects/list.rb

```diff
@@ -0,0 +1,239 @@
+require "y2storage/proposal/space_maker_prospects/resize_partition"
+require "y2storage/proposal/space_maker_prospects/wipe_disk"
+
+module Y2Storage
+  module Proposal
+    module SpaceMakerProspects
+      # A set of prospect actions SpaceMaker can perform to reach its goal
+      class List
+        include Yast::Logger
+
+        # @param settings [ProposalSettings] see {#settings}
+        # @param disk_analyzer [DiskAnalyzer] see {#analyzer}
+        def initialize(settings, disk_analyzer)
+          @settings = settings
+          @analyzer = disk_analyzer
+
+          @all_delete_partition_entries = {
+            linux:   [],
+            windows: [],
+            other:   []
+          }
+
+          @resize_partition_without_linux_entries = []
+          @resize_partition_with_linux_entries = []
+          @wipe_disk_entries = []
+        end
+
+        # Adds to the set all the prospect actions for the given disk
+        #
+        # @param disk [Disk] disk to act upon
+        # @param lvm_helper [Proposal::LvmHelper] contains information about the
+        #     planned LVM logical volumes and how to make space for them
+        # @param keep [Array<Integer>] sids of partitions that should not be deleted
+        def add_entries(disk, lvm_helper, keep = [])
+          add_delete_partition_entries(disk, keep)
+          add_resize_entries(disk)
+          add_wipe_entries(disk, lvm_helper)
+        end
+
+        # Adds to the set all the prospect actions about deleting partitions of
+        # the given disk (i.e. entries of type {SpaceMakerProspects::DeletePartition})
+        #
+        # @param disk [Disk] disk to act upon
+        # @param keep [Array<Integer>] sids of partitions that should not be deleted
+        def add_delete_partition_entries(disk, keep = [])
+          entries = delete_prospects_for_disk(disk, keep)
+          linux, non_linux = entries.partition { |e| e.partition_type == :linux }
+          windows, other = non_linux.partition { |e| e.partition_type == :windows }
+
+          delete_partition_entries(:linux).concat(linux.sort_by(&:region_start).reverse)
+          delete_partition_entries(:windows).concat(windows.sort_by(&:region_start).reverse)
+          delete_partition_entries(:other).concat(other.sort_by(&:region_start).reverse)
+        end
+
+        # Entries of type #{SpaceMakerProspects::DeletePartition}
+        #
+        # @param type [Symbol, nil] optional type to filter the result
+        def delete_partition_entries(type = nil)
+          if type.nil?
+            @all_delete_partition_entries.values.flatten
+          else
+            @all_delete_partition_entries[type]
+          end
+        end
+
+        # Next prospect action that should be executed by SpaceMaker
+        #
+        # @return [SpaceMakerProspects::Base, nil] nil if there are no more
+        #   available prospects
+        def next_available_entry
```

Comment:
Maybe, it would be clearer an implementation like:

```suggestion
        def next_available_entry
          next_resize_partition(allow_linux_in_disk: false) ||
            next_delete_partition(allow_windows: false) ||
            next_wipe_disk ||
            next_resize_partition(allow_linux_in_disk: true) ||
            next_delete_partition(allow_windows: true)
        end
```

Reply:
Hmmm, maybe. But the `:allow_windows` name is already imprecise. It actually excludes Windows systems AND partitions marked as `#last_resort?`...

So I would keep it as it is for the time being. Trying to shorten it too much can lead to a lot of new argument in the methods with not-so-precise names.